src/controller/mod.rs

```diff
@@ -518,6 +519,23 @@ impl ControllerInner {
         r
     }
 
+    /// Initiaties log recovery by sending a
+    /// StartRecovery packet to each base node domain.
+    pub fn recover(&mut self, _: ()) {
+        info!(self.log, "Recovering from log");
+        for (_name, index) in self.inputs(()).iter() {
+            let node = &self.ingredients[*index];
+            let domain = self.domains.get_mut(&node.domain()).unwrap();
+            domain.send(box payload::Packet::StartRecovery).unwrap();
+        }
+
+        for (_name, index) in self.inputs(()).iter() {
+            let node = &self.ingredients[*index];
+            let domain = self.domains.get_mut(&node.domain()).unwrap();
+            domain.wait_for_ack().unwrap();
+        }
```

Comment:
I'm pretty sure this can deadlock. Imagine two domains that both have egresses to the other. Both are told to recover at the same time. If there's enough data, they will eventually fill up the channels between one another, and deadlock.

I think you'll have to wait for an ack from each node recovery sequentially unfortunately. At least until we can come up with a cleverer scheme.

Reply:
That's unfortunate :/ changed back to sequential recovery in https://github.com/mit-pdos/distributary/pull/37/commits/8126b20de628f3f86d1b863ceca7f30128efb7ed.

Reply:
I agree that sequential recovery is the safe option. Thinking about it, though, I don't think there is (currently) any way that the described deadlock would occur: it would require that a new domain adds a node into an existing domain 
 *below* its own nodes: i.e., an existing `A -> B` relation turns into either `A -> B -> A` (not permitted) or independent `A -> B`, `B -> A` paths (which require `B` to add an edge, and thus a node, into `A`). I don't think our current assignment logic can ever do this?

Reply:
I'm hesitant to say that the current logic can't do that actually. It does funky things when there are joins (which I think essentially ends up as placing the join in the same domain as one of its ancestors).